Three json

Hoodgail opened the pull request:
- [x] Object3D
  - [x] Scene
  - [x] Mesh
    - [x] InstancedMesh
  - [x] Line
    - [x] LineLoop
    - [x] LineSegments
  - [x] Skeleton
  - [x] Camera 
    - [x] PerspectiveCamera
    - [x] OrthographicCamera
    - [x] CubeCamera
    - [x] ArrayCamera
  - [x] Group
  - [x] Bone
  - [x] Sprite
  - [x] Points
  - [x] LOD
  - [ ] Light
    - [x] HemisphereLight
    - [ ] SpotLight
    - [x] PointLight
    - [x] DirectionalLight
    - [x] AmbientLight
    - [ ] RectAreaLight
    - [ ] LightProbe 
  - [ ] Audio
    - [ ] PositionalAudio 
  - [ ] AudioListener

- [ ] FogBase
  - [ ] Fog
  - [ ] FogExp2

- [x] Math `(Core)`
  - [x] Euler
  - [x] Quaternion
  - [x] AnimationClip
  - [x] KeyframeTrack 
  - [x] Matrix3
  - [x] Matrix4
  - [x] Vector2
  - [x] Curve
    - [x] CurvePath
      - [x] Path

- [x] Texture
  - [ ] VideoTexture
  - [ ] DataTexture
  - [ ] CompressedTexture
  - [ ] CubeTexture
  - [ ] CanvasTexture
  - [ ] DepthTexture

- [x] Source 
- [x] BufferGeometry
- [x] BufferAttribute

- [x] KeyframeTrack
  - [x] BooleanKeyframeTrack
  - [x] ColorKeyframeTrack
  - [x] NumberKeyframeTrack
  - [x] QuaternionKeyframeTrack
  - [x] StringKeyframeTrack
  - [x] VectorKeyframeTrack

**Finalization**
- [ ] Implement the respective json interfaces to their toJSON functions.
- [ ] Revise

Hoodgail committed:
Add Skeleton.toJSON

Addding a  Skeleton.toJSON

Hoodgail committed:
Update Skeleton.d.ts

Added Skeleton.fromJson
https://github.com/mrdoob/three.js/blob/0fbae6f682f6e13dd9eb8acde02e4f50c0b73935/src/objects/Skeleton.js#L221

Hoodgail committed:
Update Skeleton.d.ts

Optional parameter 
https://github.com/mrdoob/three.js/blob/0fbae6f682f6e13dd9eb8acde02e4f50c0b73935/src/objects/Skeleton.js#L15

Hoodgail committed:
Update Skeleton.d.ts

Hoodgail committed:
Update Skeleton.d.ts

Hoodgail committed:
BufferAttributeJSON

JSON data for BufferAttribute

Methuselah96 committed:
Format

Hoodgail committed:
KeyframeTrackJSON

Hoodgail committed:
AnimationClipJSON, KeyframeTrackJSON

Hoodgail committed:
QuaternionTuple

Hoodgail committed:
EulerTuple

Hoodgail committed:
Merge remote-tracking branch 'origin/patch-6' into three-json

Hoodgail committed:
SkeletonJSON

Hoodgail committed:
Merge branch 'dev' into three-json

Hoodgail committed:
Vector2Tuple

Hoodgail committed:
Matrix3Tuple

Hoodgail committed:
CurveJSON

Hoodgail committed:
Merge branch 'three-json' of https://github.com/Hoodgail/three-ts-types into three-json

Hoodgail committed:
SourceJSON

Hoodgail committed:
CurvePathJSON

Hoodgail committed:
BufferGeometryJSON

Hoodgail committed:
TextureJSON

Hoodgail committed:
PathJSON

Hoodgail committed:
Object3DJSON

Hoodgail committed:
Prettier & Lint Fixes

Hoodgail commented:
@Methuselah96 What do you think of the progress so far?
let me know if how i'm doing it is good before i continue

Methuselah96 commented on types/three/src/Meta.d.ts:
```diff
@@ -0,0 +1,5 @@
+    version: Version;
+    type: Type;
+    generator: Generator;
+}
```
I would really like to avoid creating a file that doesn't exist in three.js. Looks likes this is used in 5 places. Could we either just in-line it in the places it's used (and get rid of the interface altogether) or move it into a file that already exists like `Object3D.d.ts`?

Methuselah96 commented on types/three/src/animation/KeyframeTrack.d.ts:
```diff
@@ -4,6 +4,18 @@ import { CubicInterpolant } from './../math/interpolants/CubicInterpolant';
 import { InterpolationModes } from '../constants';
 
+export interface KeyframeTrackJSON<Type extends string = 'KeyframeTrack'> {
+    type: Type;
```
Can you help me understand how the `type` property works and why it's generic in all of them?

Hoodgail replied:
> Can you help me understand how the type property works and why it's generic in all of them?

Well first, i added it because i thought the type and generator in the Meta changed based on the type. but i was wrong and just kept it generic on all of them.

But it might be good for extended objects, like bone which is just Object3D with the .type as "Bone" 
I was thinking it would be better than overriding the type property, like Group

> I would really like to avoid creating a file that doesn't exist in three.js. Looks likes this is used in 5 places. Could we either just in-line it in the places it's used (and get rid of the interface altogether) or move it into a file that already exists like Object3D.d.ts?

Yeah i was thinking the same thing, i just didnt know what to do, i'll move it to `Object3D.d.ts`

I gave the meta its own interface just incase it changes, we woulnt have to change it in everything.
Since its pretty much the same, keeps from stuff being repetitive

Methuselah96 reviewed (commented):
Looking good overall, thanks for working on this.

Hoodgail committed:
BoneJSON

Hoodgail committed:
GroupJSON

Hoodgail committed:
MeshJSON, InstancedMeshJSON

Hoodgail committed:
SkinnedMeshJSON

Hoodgail committed:
Format

Hoodgail commented:
@kourser Reminder to finish this.

Hoodgail committed:
Merge branch 'master' into three-json

Hoodgail committed:
Format

Hoodgail committed:
Scene & Fog

Hoodgail commented:
*Note to self*
https://github.com/three-types/three-ts-types/blob/8109927fa28df9359537ea3cdc1b4c777b905a12/types/three/src/core/Object3D.d.ts#L646

expected output format:
```js
{
    geometries: geojson[]
    materials: materialjson[]
    textures: texturejson[]
    images: imagejson[]
    object: Object3DJSON
}
```

Hoodgail committed:
Camera

Hoodgail committed:
PerspectiveCamera

Hoodgail committed:
OrthographicCamera

Hoodgail committed:
Format

Hoodgail committed:
Remove Meta.d.ts
https://github.com/three-types/three-ts-types/pull/426/files/3363b9eaab245bdde33db8c6ab76ec179c4c3a0c#r1178560510

Hoodgail committed:
Fix
Error:   41:18  error  Array type using 'Array<GeometryGroup>' is forbidden for simple types. Use 'GeometryGroup[]' instead  @typescript-eslint/array-type
Error:   3:18  error  An interface declaring no members is equivalent to its supertype  @typescript-eslint/no-empty-interface
Error:   3:18  error  An interface declaring no members is equivalent to its supertype  @typescript-eslint/no-empty-interface

Hoodgail committed:
Line

Hoodgail committed:
LineLoop

Hoodgail committed:
LineSegments

Hoodgail committed:
CubeCamera

Hoodgail committed:
Sprite

Hoodgail committed:
Points

Hoodgail committed:
BatchedMesh

Hoodgail committed:
LOD

Hoodgail committed:
Minor fixes

Hoodgail committed:
Light & LightShadow

Hoodgail committed:
HemisphereLight

Hoodgail committed:
SpotLight

Hoodgail committed:
PointLight

Hoodgail committed:
AmbientLight

Hoodgail committed:
Light, Dir light & Light shadow

Hoodgail committed:
Format

Hoodgail committed:
Typo

Hoodgail committed:
Fixx:   9:6  error  All declarations in this module are exported automatically. Prefer to explicitly write 'export' for clarity. If you have a good reason not to export this declaration, add 'export {}' to the module to shut off automatic exporting  @definitelytyped/strict-export-declare-modifier

Methuselah96 committed:
Remove package-lock.json

Methuselah96 committed:
Update Object3D

Methuselah96 committed:
Animation

Methuselah96 committed:
Cameras

Methuselah96 committed:
BufferGeometry

Methuselah96 committed:
Curves

Methuselah96 committed:
Lights

Methuselah96 committed:
Objects

Methuselah96 committed:
Scenes

Methuselah96 committed:
Textures

Methuselah96 committed:
Add examples

Methuselah96 committed:
Update patch

Methuselah96 committed:
Delete examples

Methuselah96 committed:
Update declaration

Methuselah96 reviewed (approved):
@Hoodgail Thanks for your work on this! I went ahead and did some cleanup of some of the changes, hopefully while preserving the intent of your changes. Let me know if I messed anything up and feel free to keep building on this with more PRs, I just figured this would be a good place to get some of it merged.

Hoodgail commented:
Thank you

mattrossman commented:
@Hoodgail  I traced this PR as the source of the `QuaternionTuple` change here

https://github.com/three-types/three-ts-types/blame/a6fadb5785285a73ccf9f9dbc8b81186c7ebfa42/types/three/src/math/Quaternion.d.ts#L14

Since `fromArray` is typed as
```ts
fromArray(array: QuaternionTuple, offset?: number): this;
```
it produces an error when provided an array that doesn't strictly have 4 elements, which is expected when passing an `offset`. This makes it impractical to use in common scenarios such as when the array is a `Float32Array` (e.g. packed buffer of multiple quaternions) or otherwise contains >4 elements.

![CleanShot 2024-09-17 at 18 03 09@2x](https://github.com/user-attachments/assets/51c0243d-0021-42cd-bdda-42172cb31c56)

https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAKjgQwM5wCoAkBKBRPOAMyghDgHIYALKAUzooCgmBjCAO1XgEcBXZPAC8cDnQDumXAQB0ARQEw6UDsE4AKAJQt2XeDACeYOgBMAglCjIDcEWMkAxADYRBAZgBMFqwfUAObV1uOAgiIlQ6YTgAFhZ+QRkSMm9rdUNjc0trABoQsIiYTSA

Is there a reason `array` needs to be typed this way?

Hoodgail commented:
@mattrossman It looks like this was an oversight on my part. I had completely forgot about the offset argument

It should follow the same way its written in Vector3 at https://github.com/three-types/three-ts-types/blob/a6fadb5785285a73ccf9f9dbc8b81186c7ebfa42/types/three/src/math/Vector3.d.ts#L269
![{D400AF8F-C00C-4E49-B28E-30EA41E23447}](https://github.com/user-attachments/assets/0acebcba-cc25-44a5-9da9-4f47f744fe7a)

Will you make a pull request? or should i? My apologies !

mattrossman commented:
@Hoodgail I've submitted a PR https://github.com/three-types/three-ts-types/pull/1230, thank you for the suggested fix!
